pkg/statsd/receiver.go

```diff
@@ -96,99 +83,24 @@ func (mr *MetricReceiver) Receive(ctx context.Context, c net.PacketConn) {
 			continue
 		}
 		// TODO consider updating counter for every N-th iteration to reduce contention
-		atomic.AddUint64(&mr.packetsReceived, uint64(packetCount))
-		atomic.StoreInt64(&mr.lastPacket, time.Now().UnixNano())
-		atomic.AddUint64(&mr.batchesRead, 1)
+		atomic.AddUint64(&dr.packetsReceived, uint64(packetCount))
+		atomic.StoreInt64(&dr.lastPacket, time.Now().UnixNano())
+		atomic.AddUint64(&dr.batchesRead, 1)
 		for i := 0; i < packetCount; i++ {
 			addr := messages[i].Addr
 			nbytes := messages[i].N
 			buf := messages[i].Buffers[0]
-			if err := mr.handlePacket(ctx, addr, buf[:nbytes]); err != nil {
-				if err == context.Canceled || err == context.DeadlineExceeded {
-					return
-				}
-				log.Warnf("Failed to handle packet: %v", err)
+			msg := make([]byte, nbytes)
+			copy(msg, buf)
+			dg := &Datagram{
+				IP:  getIP(addr),
+				Msg: msg,
 			}
+			dr.out <- dg
```

Comment:
For code below I see that the shutdown order between parser/receiver is not defined (they are started within same Stage). That means this can block forever if nobody is reading (and/or the channel is full). I think it should be a `select` with `<-ctx.Done()`.

Reply:
Yeah good catch. I'll make the change to check our context isn't cancelled. 

Would you also recommend splitting off into a new stage(s)? I'm curious about the triggers for creating a new stage... why doesn't the flusher run within it's own stage? Should it?

If I'm not mistaken is it just sensible to run each worker in it's own stage so that shutdown matches startup?

Reply:
Stages are necessary to ensure ordered shutdown. If order does not matter then shutdown within a stage will happen concurrently. My rule of thumb is to only create stages when order matters. This also serves as documentation - it is done that way because there is a reason for it.
Feel free to keep it as it is or change it back - I'd like to see if it is possible to get rid of context checking in some places and instead rely on the data channel being closed when producer will no longer send stuff. This is more performant that `select` and reduces boilerplate. But that is for later.